plutus-core/untyped-plutus-core/src/UntypedPlutusCore/Evaluation/Machine/SteppableCek/Internal.hs

```diff
@@ -171,27 +178,28 @@ returnCek NoFrame val = do
 -- s , {_ A} ◅ abs α M  ↦  s ; ρ ▻ M [ α / A ]*
 returnCek (FrameForce _ ctx) fun = forceEvaluate ctx fun
 -- s , [_ (M,ρ)] ◅ V  ↦  s , [V _] ; ρ ▻ M
-returnCek (FrameApplyArg _funAnn argVarEnv arg ctx) fun =
+returnCek (FrameAwaitFunTerm _funAnn argVarEnv arg ctx) fun =
     -- MAYBE: perhaps it is worth here to merge the _funAnn with argAnn
-    pure $ Computing (FrameApplyFun (termAnn arg) fun ctx) argVarEnv arg
+    pure $ Computing (FrameAwaitArg (termAnn arg) fun ctx) argVarEnv arg
 -- s , [(lam x (M,ρ)) _] ◅ V  ↦  s ; ρ [ x  ↦  V ] ▻ M
 -- FIXME: add rule for VBuiltin once it's in the specification.
-returnCek (FrameApplyFun _ fun ctx) arg =
+returnCek (FrameAwaitArg _ fun ctx) arg =
     applyEvaluate ctx fun arg
 -- s , [_ V1 .. Vn] ◅ lam x (M,ρ)  ↦  s , [_ V2 .. Vn]; ρ [ x  ↦  V1 ] ▻ M
-returnCek (FrameApplyValues ann args ctx) fun = case args of
-    (arg:rest) -> applyEvaluate (FrameApplyValues ann rest ctx) fun arg
-    _          -> returnCek ctx fun
+returnCek (FrameAwaitFunValue _ arg ctx) fun =
+    applyEvaluate ctx fun arg
 -- s , constr I V0 ... Vj-1 _ (Tj+1 ... Tn, ρ) ◅ Vj  ↦  s , constr i V0 ... Vj _ (Tj+2... Tn, ρ)  ; ρ ▻ Tj+1
 returnCek (FrameConstr ann env i todo done ctx) e = do
-    let done' = done `DList.snoc` e
+    let done' = ConsStack e done
     case todo of
         (next : todo') -> computeCek (FrameConstr ann env i todo' done' ctx) env next
-        _              -> returnCek ctx $ VConstr i (toList done')
+        _              -> returnCek ctx $ VConstr i done'
 -- s , case _ (C0 ... CN, ρ) ◅ constr i V1 .. Vm  ↦  s , [_ V1 ... Vm] ; ρ ▻ Ci
 returnCek (FrameCases ann env cs ctx) e = case e of
     (VConstr i args) -> case cs ^? wix i of
-        Just t  -> computeCek (FrameApplyValues ann args ctx) env t
+        Just t  ->
+              let ctx' = transferArgStack ann args ctx
+              in computeCek ctx' env t
```

Comment:
This `let` is the only difference between this PR and my original one that I'm able to spot. I can't imagine it making this much difference, but perhaps still is worth trying to inline it or to mark as `INLINE`.